File: Data Structures & Algorithms/binary-tree-diameter/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def checkLength(self, root, res):
        if root == None:
            return 0, 0
        # Search left depth
        left, res1 = self.checkLength(root.left, res)
        # Search right depth
        right, res2 = self.checkLength(root.right, res)
        # assign node val of max num edges
        count = max(left, right)
        print(1 + count)
        res = max(res1, res2, left + right)
        # Send up
        return 1 + count, res
        
        

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        _, res = self.checkLength(root, res)
        return res
        

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:

    bool res;

    int checkDepth(TreeNode* root){
        if(root == nullptr){
            return 0;
        }

        int left = this->checkDepth(root->left);
        int right = this->checkDepth(root->right);

        if(abs(left - right) > 1){
            this->res = false;
        }

        return max(left, right)+1;
    }

    
    bool isBalanced(TreeNode* root) {
        res = true;
        int count = this->checkDepth(root);
        return res;
    }
};
